frontend/src/pages/new_question/queryBuilder.js

```diff
@@ -0,0 +1,226 @@
+    const { isValid, newRoot } = queryBuilderUtils.isValidGraph(trimmedQueryGraph, keptRoot);
+    if (isValid) {
+      updateQueryGraph(trimmedQueryGraph);
+      setRootNode(newRoot);
+    } else {
+      displayAlert('error', 'You cannot delete this connection. It would make this query invalid.');
+    }
+  }
+
+  /**
+   * Update an edge in the query graph
+   * @param {string} edgeId edge key
+   * @param {string} edgeType subject or object of edge
+   * @param {string} nodeId node id
+   * @returns boolean
+   */
+  function updateEdge(edgeId, edgeType, nodeId) {
+    const clonedQueryGraph = _.cloneDeep(query_graph);
+    if (!nodeId) {
+      const newNodeId = queryBuilderUtils.getNextNodeID(clonedQueryGraph);
+      clonedQueryGraph.nodes[newNodeId] = defaultNode();
+      clonedQueryGraph.edges[edgeId][edgeType] = newNodeId;
+    } else {
+      clonedQueryGraph.edges[edgeId][edgeType] = nodeId;
+    }
+    const keptRoot = queryBuilderUtils.computeRootNode(clonedQueryGraph, rootNode);
+    const trimmedQueryGraph = queryBuilderUtils.trimDetached(clonedQueryGraph, keptRoot);
+    const { isValid, newRoot } = queryBuilderUtils.isValidGraph(trimmedQueryGraph, keptRoot);
+    if (isValid) {
+      updateQueryGraph(trimmedQueryGraph);
+      setRootNode(newRoot);
+      return true;
+    }
+    displayAlert('error', 'You cannot change this connection. It would make this query invalid.');
+    return false;
+  }
+
+  /**
+   * Update an edge predicate
+   * @param {string} edgeId edge id in the query graph edges
+   * @param {array} predicateValue an array containing selected edge predicates
+   */
+  function updateEdgePredicate(edgeId, predicateValue) {
+    const clonedQueryGraph = _.cloneDeep(query_graph);
+    clonedQueryGraph.edges[edgeId].predicate = predicateValue;
+    updateQueryGraph(clonedQueryGraph);
+  }
+
+  /**
+   * Update an existing node
+   * @param {string} nodeId node id
+   * @param {object|null} updatedNode updated node object
+   */
+  function updateNode(nodeId, updatedNode) {
+    const clonedQueryGraph = _.cloneDeep(query_graph);
+    clonedQueryGraph.nodes[nodeId] = updatedNode || defaultNode();
+    updateQueryGraph(clonedQueryGraph);
+  }
+
+  /**
+   * Delete an existing node and then trim the graph
+   * @param {string} nodeId node id
+   */
+  function deleteNode(nodeId) {
+    const clonedQueryGraph = _.cloneDeep(query_graph);
+    delete clonedQueryGraph.nodes[nodeId];
+    const trimmedQueryGraph = queryBuilderUtils.trimDetachedEdges(clonedQueryGraph, nodeId, rootNode);
+    const { isValid, newRoot } = queryBuilderUtils.isValidGraph(trimmedQueryGraph, rootNode);
+    if (isValid) {
+      updateQueryGraph(trimmedQueryGraph);
+      setRootNode(newRoot);
+    } else {
+      displayAlert('error', 'You cannot delete this term. It would make this query invalid.');
+    }
+  }
+
+  /**
+   * Edge list with the first edge with the root node as its subject first
```

Comment:
It isn't always, but for a sentence structure we want to try and put the root on the left side. I added some logic here 3e90b6761480f74dce17323c099f35c6442ee08c in case the root is not an edge subject.